### ISSUE TITLE
[aws] Native CloudWatch logs plugin segfaults when timestamps differ too greatly

### ISSUE THREAD

**Issue by ashlineldridge** [fluent/fluent-bit#2612]
## Bug Report

**Describe the bug**
The native C CloudWatch logs plugin segfaults when two consecutive log records have timestamps that differ too greatly. I can consistently reproduce this bug with the log file provided below.

**To Reproduce**

Log file named `test.log`:

```
{"log":"hello","time":"2020-09-14T05:03:39.880638969Z"}
{"log":"goodbye","time":"2020-09-18T05:31:18.850360396Z"}
```

Configuration file named `test.conf`:

```
[SERVICE]
    Flush 5
    Daemon Off
    Log_Level info
    Parsers_File /fluent-bit/etc/parsers.conf
    HTTP_Server On
    HTTP_Listen 0.0.0.0
    HTTP_Port 2020

[INPUT]
    Name tail
    Path test.log
    Parser docker
    Tag kube.*
    Mem_Buf_Limit 64MB
    Skip_Long_Lines On

[OUTPUT]
    Name cloudwatch_logs
    Match kube.*
    region ap-southeast-2
    log_group_name /test/application
    log_stream_prefix application.
    auto_create_group true
```

Authenticate with AWS (to produce a credentials file under `~/.aws`) and then run the following in the same directory as `test.log` and `test.conf`:

```
docker run -ti -v "${PWD}:/src" -v "${HOME}/.aws:/root/.aws" -w /src fluent/fluent-bit:1.5 /fluent-bit/bin/fluent-bit -c /src/test.conf
```

Produces the following error:

```
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/09/29 04:00:45] [ info] [engine] started (pid=1)
[2020/09/29 04:00:45] [ info] [storage] version=1.0.5, initializing...
[2020/09/29 04:00:45] [ info] [storage] in-memory
[2020/09/29 04:00:45] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/09/29 04:00:45] [ info] [http_server] listen iface=0.0.0.0 tcp_port=2020
[2020/09/29 04:00:45] [ info] [sp] stream processor started
[2020/09/29 04:00:45] [ info] inotify_fs_add(): inode=16119175 watch_fd=1 name=test.log
[2020/09/29 04:00:49] [ info] [output:cloudwatch_logs:cloudwatch_logs.0] Creating log group /test/application
[2020/09/29 04:00:49] [ info] [output:cloudwatch_logs:cloudwatch_logs.0] Log Group /test/application already exists
[2020/09/29 04:00:49] [ info] [output:cloudwatch_logs:cloudwatch_logs.0] Creating log stream application.kube.test.log in log group /test/application
[engine] caught signal (SIGSEGV)
[2020/09/29 04:00:49] [ info] [output:cloudwatch_logs:cloudwatch_logs.0] Log Stream application.kube.test.log already exists
#0  0x5608688d55d8      in  compare_events() at plugins/out_cloudwatch_logs/cloudwatch_api.c:159
#1  0x7faa61e28a8b      in  ???() at ???:0
#2  0x7faa61e293b5      in  ???() at ???:0
#3  0x5608688d604e      in  send_log_events() at plugins/out_cloudwatch_logs/cloudwatch_api.c:466
#4  0x5608688d65ff      in  add_event() at plugins/out_cloudwatch_logs/cloudwatch_api.c:573
#5  0x5608688d6992      in  process_and_send() at plugins/out_cloudwatch_logs/cloudwatch_api.c:674
#6  0x5608688d4d31      in  cb_cloudwatch_flush() at plugins/out_cloudwatch_logs/cloudwatch_logs.c:362
#7  0x56086884c50a      in  output_pre_cb_flush() at include/fluent-bit/flb_output.h:449
#8  0x560868c80346      in  co_init() at lib/monkey/deps/flb_libco/amd64.c:117
#9  0xffffffffffffffff  in  ???() at ???:0
```

If I adjust the log record timestamps so that they are closer together then no segfault occurs.

**Expected behavior**
The application should not crash and deal with any errors by logging them.

**Your Environment**
* Version used: 1.5.7
* Configuration: Error can be reproduced locally and on Kubernetes
* Environment name and version: Local Docker container and Kubernetes 1.16
* Filters and plugins: Native CloudWatch output plugin

**Additional context**
In our Kubernetes environments this causes some of the Fluent Bit pods to enter into a crash-backoff loop which results in no more logs from the affected nodes.

**Comment by ashlineldridge** [comment]
The bug seems to be related to Fluent Bit's handling of CloudWatch's [tooOldLogEventEndIndex](https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_RejectedLogEventsInfo.html) response.

The external CloudWatch Golang plugin doesn't suffer from the same issue (below command requires changing `Name cloudwatch_logs` to `Name cloudwatch` in `test.conf`).

```
> docker run -ti -v "${PWD}:/src" -v "${HOME}/.aws:/root/.aws" -w /src amazon/aws-for-fluent-bit:2.7.0 /fluent-bit/bin/fluent-bit -e /fluent-bit/cloudwatch.so -c test.conf

Fluent Bit v1.5.6
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/09/29 04:28:33] [ info] [engine] started (pid=1)
[2020/09/29 04:28:33] [ info] [storage] version=1.0.5, initializing...
[2020/09/29 04:28:33] [ info] [storage] in-memory
[2020/09/29 04:28:33] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
INFO[0000] [cloudwatch 0] plugin parameter log_group = '/test/application'
INFO[0000] [cloudwatch 0] plugin parameter log_stream_prefix = 'application.'
INFO[0000] [cloudwatch 0] plugin parameter log_stream_name = ''
INFO[0000] [cloudwatch 0] plugin parameter region = 'ap-southeast-2'
INFO[0000] [cloudwatch 0] plugin parameter log_key = ''
INFO[0000] [cloudwatch 0] plugin parameter role_arn = ''
INFO[0000] [cloudwatch 0] plugin parameter new_log_group_tags = ''
INFO[0000] [cloudwatch 0] plugin parameter log_retention_days = '0'
INFO[0000] [cloudwatch 0] plugin parameter endpoint = ''
INFO[0000] [cloudwatch 0] plugin parameter sts_endpoint = ''
INFO[0000] [cloudwatch 0] plugin parameter credentials_endpoint =
INFO[0000] [cloudwatch 0] plugin parameter log_format = ''
[2020/09/29 04:28:33] [ info] [http_server] listen iface=0.0.0.0 tcp_port=2020
[2020/09/29 04:28:33] [ info] [sp] stream processor started
[2020/09/29 04:28:33] [ info] inotify_fs_add(): inode=16119175 watch_fd=1 name=test.log
INFO[0004] [cloudwatch 0] Log group /test/application already exists
WARN[0005] [cloudwatch 0] 1 log events were marked as too old by CloudWatch
```

**Comment by PettitWesley** [comment]
@ashlineldridge Thank you for this detailed bug report!

With tag v1.5.7, I am able to repro. However, with the code in master I am not. I recently discovered some issues with the logic for payload creation which I fixed here: https://github.com/fluent/fluent-bit/commit/0740e09b6399a3ed8f85c1d497cd3f7d145601b1

I believe your bug report is a sub-case of the larger issue that I was fixing. I believe the next planned release is 1.6.0 and it will contain the fix.

@edsiper
